cmd/gtfTools/filter.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/gtf"
+	"log"
+	"os"
+)
+
+// FilterSettings defines the usage settings for the gtfTools filter subcommand.
+type FilterSettings struct {
+	InFile       string
+	OutFile      string
+	GeneNameList string
+}
+
+func filterUsage(filterFlags *flag.FlagSet) {
+	fmt.Printf("gtfTools filter - a tool for filtering GTF records.\n" +
+		"Usage:\n" +
+		"gtfTools filter in.gtf out.gtf\n" +
+		"options:\n")
+	filterFlags.PrintDefaults()
+}
+
+// parseFilterArgs is the main function for the gtfTools filter subcommand. It parses options
+// and executes the function gtfFilter.
+func parseFilterArgs() {
+	var expectedNumArgs int = 2
+	var err error
+	filterFlags := flag.NewFlagSet("filter", flag.ExitOnError)
+	var geneNameList *string = filterFlags.String("geneNameList", "", "Specify a new-line delimited file containing the geneNames for records to be retained.\n")
+	err = filterFlags.Parse(os.Args[2:])
+	exception.PanicOnErr(err)
+	filterFlags.Usage = func() { filterUsage(filterFlags) }
+
+	if len(filterFlags.Args()) != expectedNumArgs {
+		filterFlags.Usage()
+		log.Fatalf("Error: expecting %d arguments, but got %d\n",
+			expectedNumArgs, len(filterFlags.Args()))
+	}
+
+	inFile := filterFlags.Arg(0)
+	outFile := filterFlags.Arg(1)
+
+	s := FilterSettings{
+		InFile:       inFile,
+		OutFile:      outFile,
+		GeneNameList: *geneNameList,
+	}
+	gtfFilter(s)
+}
+
+func gtfFilter(s FilterSettings) {
+	var pass, foundInMap bool
+	var err error
+	var geneNameMap = make(map[string]bool)
+	out := fileio.EasyCreate(s.OutFile)
+
+	if s.GeneNameList != "" {
+		geneNames := fileio.Read(s.GeneNameList)
+		for i := range geneNames {
+			geneNameMap[geneNames[i]] = true
+		}
+	}
+
+	records := gtf.Read(s.InFile)
+	for currGene := range records {
+		pass = true
+		if s.GeneNameList != "" {
+			if _, foundInMap = geneNameMap[records[currGene].GeneName]; !foundInMap {
+				pass = false
+			}
+		}
+		if pass {
+			gtf.WriteToFileHandle(out, records[currGene])
```

Comment:
Note to self: records[currGene] or just currGene? records[currGene] is correct.
records is the output of gtf.Read so the type is map[string]*Gene.
currGene is the variable that ranges through records, and represents the key of the map.
This is why records[currGene] is correct.